Feature/frontend integration

AlexBud369 opened the pull request:
#### Features Implemented

**Core Setup:**
- Initialized project with Vite + React + TypeScript + SWC
- Configured Tailwind CSS and Vite (via `tailwind.config.js`, `vite.config.ts`)
- Added global styles (`index.css`) and TypeScript interfaces (`index.ts`)

**State & API:**
- Zustand stores (`authStore`, `loadingStore`) for authentication and loading states
- Axios-based API service (`api.ts`) with interceptors for token refresh and loading states
- Services for auth, users, books, and tags (`authApi`, `bookApi`, etc.)

**Components:**
- Reusable UI: `Button`, `ErrorBoundary`, `BookCard`, `BookForm` (with react-hook-form + yup validation)
- Layout: `Header` (navigation, language switcher), `PrivateRoute` (auth/admin guards)
- Forms: `LoginForm`, `RegisterForm`, `BookForm` (responsive, validated)

**Pages:**
- Auth: `LoginPage`, `RegisterPage`, `UserProfilePage`
- Books: `BookListPage` (filtering/pagination), `BookDetailsPage`, `CreateBookPage`, `EditBookPage`
- Admin: `AdminPanel` for user/book management

**Internationalization:**
- i18n setup (`i18n.ts`) with EN/RU translations for UI/error messages

**Styling & UX:**
- Responsive designs (e.g., `Header`, `BookForm` with breakpoints)
- Fixed duplicate toast notifications and layout issues

AlexBud369 committed:
chore: initial frontend setup with Vite, React, and TypeScript + SWC

AlexBud369 committed:
chore: add constants.ts file to centralize configuration values such as max lengths for title, author, description, username, email, and pagination sizes.

AlexBud369 committed:
feat: add BookCard component with styled book display
- Implement BookCard component to display book details with cover image, title, author, description, and tags

AlexBud369 committed:
feat: add BookForm component for book creation/editing
- Create BookForm component to handle book creation and editing with form validation using react-hook-form and yup

AlexBud369 committed:
feat: add reusable Button component with variant styles
- Implement Button component with primary, secondary, and danger variants using Tailwind CSS

AlexBud369 committed:
feat: add ErrorBoundary component for error handling
- Introduce ErrorBoundary component to catch and display errors in the React component tree, improving application robustness

AlexBud369 committed:
feat: add Header component with navigation and language switcher
- Implement Header component with navigation links, logout functionality, and language switcher (EN/RU)

AlexBud369 committed:
feat: add LoginForm component with validation
- Create LoginForm component for user authentication using react-hook-form and yup for validation

AlexBud369 committed:
feat: add RegisterForm component with validation
- Implement RegisterForm component for user registration with react-hook-form and yup for validation

AlexBud369 committed:
feat: add PrivateRoute component for protected routes
- Introduce PrivateRoute component to handle authentication and authorization, redirecting unauthenticated users to login and restricting admin-only routes

AlexBud369 committed:
feat: add AdminPanel page for user and book management
- Implement AdminPanel page to display and manage users and books with block and delete functionality

AlexBud369 committed:
feat: add BookDetailsPage for displaying book details
- Create BookDetailsPage to show book information, including title, author, description, tags, and cover image, with edit and delete options for authorized users

AlexBud369 committed:
feat: add BookListPage with filtering and pagination
- Implement BookListPage to display a list of books with search, tag filtering, sorting, and pagination

AlexBud369 committed:
feat: add CreateBookPage for book creation
- Introduce CreateBookPage to allow users to create new books using BookForm component

AlexBud369 committed:
feat: add EditBookPage for book editing
- Implement EditBookPage to enable editing of existing books using BookForm component

AlexBud369 committed:
feat: add HomePage with recent books display
- Create HomePage to display a list of recent books using BookCard component

AlexBud369 committed:
feat: add LoginPage with login form and registration link
- Implement LoginPage with LoginForm component and a link to the registration page

AlexBud369 committed:
feat: add RegisterPage with registration form and login link
- Create RegisterPage with RegisterForm component and a link to the login page

AlexBud369 committed:
feat: add UserProfilePage for profile management
- Implement UserProfilePage to allow users to view and update their profile information

AlexBud369 committed:
feat: add loadingStore and authStore for state management
- Implement loadingStore.ts and authStore.ts using Zustand to manage loading state and user authentication state
- Provide methods for setting loading status, user authentication, logout, and initialization, enabling seamless state management across the application.

AlexBud369 committed:
feat: add api.ts with Axios configuration and API services
- Introduce api.ts to configure Axios with interceptors for loading state and token refresh, and implement authApi, userApi, bookApi, and tagApi services for interacting with the backend

AlexBud369 committed:
feat: add i18n.ts for internationalization
-Implement i18n.ts with i18next, LanguageDetector, and translations for English and Russian
- Provide UI and error messages for books, users, authentication, and tags, enabling multilingual support across the application

AlexBud369 committed:
feat: add App.tsx with main layout and loading state
= Implement App.tsx as the main layout component with Header, ErrorBoundary, and Outlet for routing

AlexBud369 committed:
feat: add index.ts with type definitions
- Introduce index.ts to define TypeScript interfaces for books, users, tags, authentication, and pagination

AlexBud369 committed:
chore: add Tailwind and Vite configuration files

- Introduce post.config.cjs, tailwind.config.js, and vite.config.ts to configure Tailwind CSS and Vite for the project
- Set up Tailwind with content scanning and Vite with React and Tailwind plugins for efficient development and styling

AlexBud369 committed:
feat: add PrivateRoute with adminOnly support
- Implement PrivateRoute component to handle authentication and authorization, supporting adminOnly prop for admin-only routes
- Redirect unauthenticated users to login and non-admin users to home for admin routes.

AlexBud369 committed:
feat: add Header with improved text contrast and responsiveness
- Implement Header with navigation links, logout, and language switcher
-Improve link text contrast (text-gray-100, hover:text-blue-200) on bg-blue-700 and add flex-col sm:flex-row for better responsiveness from 576px

AlexBud369 committed:
feat: add BookForm with responsive styling
- Implement BookForm for creating/editing books with react-hook-form, yup validation, and react-select for tags
- Add responsive styling with p-4 sm:p-6, text-sm sm:text-base, and max-w-md sm:max-w-lg md:max-w-2xl for screens from 576px

AlexBud369 committed:
fix: add BookListPage with new layout and fixed duplicate toasts

AlexBud369 committed:
feat: add index.css with Tailwind CSS and global styles

AlexBud369 committed:
feat: add main.tsx with routing and toast notifications
- Implement main.tsx with BrowserRouter, Routes, and PrivateRoute for application navigation, including routes for Home, Books, Book Details, Login, Register, Profile, Create/Edit Book, and Admin Panel

AlexBud369 committed:
fix: correct duplicate toasts and responsive layout

AlexBud369 committed:
chore: update package.json and package-lock.json

budnikme commented on src/Frontend/src/pages/EditBookPage.tsx:
```diff
@@ -0,0 +1,54 @@
+      toast.success(t('BookUpdated'));
+      navigate(`/books/${id}`);
+    } catch (error: any) {
+      toast.error(t(error.response?.data?.errors?.[0] || 'UnknownError'));
```
Maybe create a dedicated interceptor for unknown errors, to avoid repeating try/catch blocks

AlexBud369 committed:
feat: add axios interceptor for error handling
- Implement axios interceptor in api.ts to centrally handle API errors, displaying translated toast messages for error.response.data.errors[0] or UnknownError, reducing try/catch boilerplate across components.

AlexBud369 committed:
feat: add pagination constants

AlexBud369 committed:
feat: enhance components with responsive styling and refactor error handling
- Add responsive Tailwind CSS styles (p-4 sm:p-6, text-sm sm:text-base, etc.) to BookCard, BookForm, Header, LoginForm, and RegisterForm for better mobile and desktop display
- Refactor error handling in BookForm, Header, LoginForm, and RegisterForm by removing try/catch and toast notifications, relying on api.ts interceptor
- Rename onSubmit/handleLogout to perform* functions for consistency. Update Header with bg-blue-400 and max-w-screen-xl mx-auto
- Add nav-link class to Header links.

AlexBud369 committed:
refactor: enhance pages with error handling and responsive styling
- Refactor error handling across AdminPanel, BookDetailsPage, CreateBookPage, EditBookPage, HomePage, LoginPage, RegisterPage, and UserProfilePage by removing try/catch and toast notifications, relying on api.ts interceptor
- Split logic into helper functions (e.g., updateBook, uploadCoverImage, blockUser)
- Add responsive Tailwind CSS styles (p-4 sm:p-6, text-2xl sm:text-3xl, text-sm sm:text-base) for better mobile and desktop experience
- Update AdminPanel with DEFAULT_* constants, BookDetailsPage with adjusted image sizes (max-w-md sm:max-w-lg), and UserProfilePage with adaptive form styling

budnikme commented on src/Frontend/src/components/BookForm.tsx:
```diff
@@ -55,41 +54,32 @@ export default function BookForm({ initialData, bookId, onSubmit }: BookFormProp
-    } catch (error: any) {
-      toast.error(t(error.response?.data?.errors?.[0] || 'UnknownError'));
-    }
+    const submitData: BookCreateDto | BookUpdateDto = {
```
Please create base class, like `BookDtoBase` with common properties

AlexBud369 committed:
refactor: create BookDtoBase for common book properties

- Create BookDtoBase interface in index.ts with common properties and extend BookCreateDto and BookUpdateDto from it to reduce duplication
- Update submitData in BookForm.tsx to use BookDtoBase for improved code maintainability and reduced duplication - Minor updates in CreateBookPage.tsx and EditBookPage.tsx for type consistency
